### DISCUSSION TITLE
How do I configure the default options, and remove some tags globally?

### DISCUSSION THREAD

**Discussion by aathar** [mganss/HtmlSanitizer#391] [General]
Hello
I can not get this to work. I am trying to inject the default options and only remove certain tags and attributes. For example, I am OK with all the default attributes except "href" how do I remove it globally through startup.cs?

```
`     public void ConfigureServices(IServiceCollection services)
        {
           services.Configure<HtmlSanitizerOptions>(options =>
            {
                options.AllowedTags.Remove("a");
                options.AllowedAttributes.Remove("href");
            });
            // Services
            services.AddSingleton<IHtmlSanitizer, HtmlSanitizer>(); `
```

This is not removing the "a" and "href" from the list of attributes and tags.

I simply want to configure it in startup.cs and use the same options globally across my application.

Can someone please help

**Comment by tiesont** [comment]
Assuming you're using the most current released version (7.1.542), something like this works:

```c#
services.AddSingleton<Ganss.XSS.IHtmlSanitizer, Ganss.XSS.HtmlSanitizer>(m =>
{
    var options = new Ganss.XSS.HtmlSanitizerOptions
    {
        AllowedTags = Ganss.XSS.HtmlSanitizer.DefaultAllowedTags,
        AllowedAttributes = Gans.XSS.HtmlSanitizer.DefaultAllowedAttributes
    };

    options.AllowedTags.Remove("a");
    options.AllowedAttributes.Remove("href");

    var sanitizer = new Ganss.XSS.HtmlSanitizer(options);
    return sanitizer;
});
```

Note that the `options` instance created above would only have the `AllowedTags` and `AllowedAttributes` populated - everything else will be empty, as the `HtmlSanitizerOptions` type does not initialize any values to their defaults. This is also a bit more verbose than it needs to be, to make it a bit easy to inspect with the debugger.

**Reply by aathar** [reply]
thank you so much. I wasn't aware of DefaultAllowedTags and DefaultAllowedAttributes.
I was literally spending so much time with this.

Is there a way to remove a bunch of attributes? Like RemoveMany() ? Or do I put the attributes that I want to remove in a for loop?

**Reply by tiesont** [reply]
`ISet` does not have a `RemoveRange()` method. You could probably create a `List<string>` for each property from the default values - that does have `RemoveRange`. Then you would use those lists to initialize the options object.

**Reply by tiesont** [reply]
Ah, that maybe wouldn't be all that helpful - RemoveRange wants a start and stop index of items to remove. It does have RemoveAll, which accepts a predicate. Not sure that this is all that elegant, but:

```c#
services.AddSingleton<Ganss.XSS.IHtmlSanitizer, Ganss.XSS.HtmlSanitizer>(m =>
{
    string[] tagsToRemove = { "a", "u", "i" };

    var allowedTags = new List<string>(Ganss.XSS.HtmlSanitizer.DefaultAllowedTags);
    allowedTags.RemoveAll(m => tagsToRemove.Contains(m, StringComparison.OrdinalIgnoreCase));

    var options = new Ganss.XSS.HtmlSanitizerOptions
    {
        AllowedTags = allowedTags.ToHashSet(),
        AllowedAttributes = Ganss.XSS.HtmlSanitizer.DefaultAllowedAttributes
    };

    var sanitizer = new Ganss.XSS.HtmlSanitizer(options);
    return sanitizer;
});
```

**Reply by tiesont** [reply]
> thank you so much. I wasn't aware of DefaultAllowedTags and DefaultAllowedAttributes. I was literally spending so much time with this.

I think there's a PR that will move them to a separate type, but yeah, they're static properties of the HtmlSanitizer class.

**Reply by mganss** [reply]
The PR has been merged and will be in 8.0. Currently, to remove a set of tags from the default set you can also do:

```c#
var options = new HtmlSanitizerOptions
{
    AllowedTags = HtmlSanitizer.DefaultAllowedTags.Except(new[] { "a", "u", "i" }).ToHashSet(),
};
```

**Reply by tiesont** [reply]
I always forget about Except 😆

**Reply by tiesont** [reply]
Just as a follow-up, the above options code would look like this in v8:

```c#
var options = new HtmlSanitizerOptions
{
    AllowedTags = HtmlSanitizerDefaults.AllowedTags.Except(new[] { "a", "u", "i" }).ToHashSet(),
};
```

@mganss Would it be worth revisiting the wiki? Think I was the last updater on most of those entries, and they're from 2018.

**Reply by mganss** [reply]
@tiesont Sure. Would you like to tackle this? IIRC we had talked about generating the API docs automatically. Perhaps we could just [link to FuGet](https://www.fuget.org/packages/HtmlSanitizer) instead?

**Reply by tiesont** [reply]
Fuget works as an API explorer, sure. I like that idea. Not sure how it'd work for the other entries.
